README.md
# Fragment Cache
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/Rarst/fragment-cache/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/Rarst/fragment-cache/?branch=master)

Fragment Cache is a WordPress plugin for partial and async caching of heavy front-end elements. It currently supports caching navigation menus, widgets, and galleries.

Caching is built on top of transients API (with enhancements provided by TLC Transients library), provides soft expiration and transparent object cache support.

# Installation

Download plugin archive from [releases section](https://github.com/Rarst/fragment-cache/releases).

Or install in plugin directory via [Composer](https://getcomposer.org/):

    composer create-project rarst/fragment-cache --no-dev

# Frequently Asked Questions

## Why fragments don't recognize logged in users / current page?

Fragment Cache implements soft expiration - when fragments expire, they are regenerated asynchronously and do not take time in front end page load. The side effect is that it is impossible to preserve context precisely and in generic way.

Fragments that must be aware of users or other context information should be excluded from caching or handled by custom implementation, that properly handles that specific context.

## How to disable caching?

### Disable handler

Caching for the fragment type can be disabled by manipulating main plugin object:

```php
global $fragment_cache;

// completely remove handler, only use before init
unset( $fragment_cache['widget'] );

// or disable handler, use after init
$fragment_cache['widget']->disable();
```

### Skip individual fragments

Caching for individual fragments can be disabled by using `fc_skip_cache` hook.

#### Widget

```php
add_filter( 'fc_skip_cache', function ( $skip, $type, $name, $args, $salt ) {

	if ( 'widget' === $type && is_a( $args['callback'][0], 'WP_Widget_Meta' ) ) {
		return true;
	}

	return $skip;
}, 10, 5 );
```

#### Menu

```php
add_filter( 'fc_skip_cache', function ( $skip, $type, $name, $args, $salt ) {

	if ( 'menu' === $type && isset( $args['name'] ) && 'Menu with login' === $args['name'] ) {
		return true;
	}

	if ( 'menu' === $type && isset( $args['theme_location'] ) && 'header' === $args['theme_location'] ) {
		return true;
	}

	return $skip;
}, 10, 5 );
```

#### Gallery

```php
add_filter( 'fc_skip_cache', function ( $skip, $type, $name, $args, $salt ) {

	if ( 'gallery' === $type && 123 === $args['post_id'] ) {
		return true;
	}

	return $skip;
}, 10, 5 );
```

# License Info

Fragment Cache own code is licensed under GPLv2+ and it makes use of code from:

 - Composer (MIT)
 - Pimple (MIT)
 - TLC Transients (GPLv2+)

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

 - This project adheres to [Semantic Versioning](http://semver.org/).
 - This change log adheres to [Keep a CHANGELOG](http://keepachangelog.com/).

## 1.3.1 - 2016-12-11

## Fixed
 - disabled cache in Customizer preview
 - implemented menus and widgets invalidation when edited in Customizer

## 1.3 - 2016-05-14 

### Added
 - `pre_wp_nav_menu` support for menu cache.

### Changed
 - Updated code style and inline documentation.
 - Changed Pimple to version 3.
 - Committed Composer lock file into project.  

### Deprecated
 - `wp_nav_menu_args` menu cache for WP < 3.9.

## 1.2 - 2014-06-29

### Added
 - [Update Blocker](https://github.com/Rarst/update-blocker) support.

## 1.1 - 2013-06-15

### Added
 - `fc_skip_cache` filter to disable cache for individual fragments.

### Changed
 - Anonymized fragment generation in front end for consistency.
 - Updated documentation. 

## 1.0 - 2013-06-08

### Added
 - Initial release.


fragment-cache.php
<?php

/*
Plugin Name: Fragment Cache
Plugin URI: https://github.com/Rarst/fragment-cache
Description: Cache generated HTML for resource intensive and time consuming components.
Author: Andrey "Rarst" Savchenko
Version: 1.3
Author URI: http://www.rarst.net/
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

namespace Rarst\Fragment_Cache;

if ( file_exists( __DIR__ . '/vendor/autoload.php' ) ) {
	require __DIR__ . '/vendor/autoload.php';
}

global $fragment_cache;

$fragment_cache = new Plugin( array(
	'timeout'       => HOUR_IN_SECONDS,
	'update_server' => new \TLC_Transient_Update_Server(),
) );

$fragment_cache->add_fragment_handler( 'menu', 'Rarst\\Fragment_Cache\\Menu_Cache' );
$fragment_cache->add_fragment_handler( 'widget', 'Rarst\\Fragment_Cache\\Widget_Cache' );
$fragment_cache->add_fragment_handler( 'gallery', 'Rarst\\Fragment_Cache\\Gallery_Cache' );

$fragment_cache->run();


php/class-plugin.php
<?php

namespace Rarst\Fragment_Cache;

use Pimple\Container;

/**
 * Main plugin's class.
 */
class Plugin extends Container {

	/** @var array $handlers Set of registered fragment handlers. */
	protected $handlers = array();

	/**
	 * Start the plugin after initial setup.
	 */
	public function run() {

		add_action( 'init', array( $this, 'init' ) );
		add_filter( 'update_blocker_blocked', array( $this, 'update_blocker_blocked' ) );
	}

	/**
	 * Enable registered fragment handlers on init.
	 */
	public function init() {

		if (
			'on' === filter_input( INPUT_POST, 'wp_customize' )
			&& empty( filter_input( INPUT_POST, 'action' ) )
			&& current_user_can( 'customize' )
		) {
			return; // We don’t want cache running in Customizer previews.
		}
		
		foreach ( $this->handlers as $key => $type ) {
			if ( isset( $this[ $type ] ) ) {
				/** @var Fragment_Cache $handler */
				$handler = $this[ $type ];
				$handler->enable();
			} else {
				unset( $this->handlers[ $key ] );
			}
		}
	}

	/**
	 * @see https://github.com/Rarst/update-blocker
	 *
	 * @param array $blocked Configuration data for blocked items.
	 *
	 * @return array
	 */
	public function update_blocker_blocked( $blocked ) {

		$blocked['plugins'][] = plugin_basename( dirname( __DIR__ ) . '/fragment-cache.php' );

		return $blocked;
	}

	/**
	 * Add (or override) cache handler and enable it.
	 *
	 * @param string $type       Handler type name.
	 * @param string $class_name Handler class name to instance.
	 */
	public function add_fragment_handler( $type, $class_name ) {

		if ( isset( $this[ $type ] ) ) {
			/** @var Fragment_Cache $handler */
			$handler = $this[ $type ];
			$handler->disable();
			unset( $this[ $type ] );
		}

		$this[ $type ] = function ( $plugin ) use ( $type, $class_name ) {
			return new $class_name( array( 'type' => $type, 'timeout' => $plugin['timeout'] ) );
		};

		if ( ! in_array( $type, $this->handlers, true ) ) {
			$this->handlers[] = $type;
		}
	}
}


php/class-widget-cache.php
<?php

namespace Rarst\Fragment_Cache;

/**
 * Cache widgets.
 */
class Widget_Cache extends Fragment_Cache {

	/**
	 * @inheritDoc
	 */
	public function enable() {

		if ( is_admin() ) {
			add_filter( 'widget_update_callback', array( $this, 'widget_update_callback' ) );
			add_action( 'wp_ajax_update-widget', array( $this, 'update_widget' ), 0 );
		} else {
			add_filter( 'widget_display_callback', array( $this, 'widget_display_callback' ), 10, 3 );
		}
	}

	/**
	 * @inheritDoc
	 */
	public function disable() {

		if ( is_admin() ) {
			remove_filter( 'widget_update_callback', array( $this, 'widget_update_callback' ) );
			remove_action( 'wp_ajax_update-widget', array( $this, 'update_widget' ), 0 );
		} else {
			remove_filter( 'widget_display_callback', array( $this, 'widget_display_callback' ), 10 );
		}
	}

	/**
	 * Adds timestamp to widget instance to use as salt.
	 *
	 * @param array $instance Widget instance to modify.
	 *
	 * @return array
	 */
	public function widget_update_callback( $instance ) {

		if ( is_array( $instance ) ) {
			$instance['fc_widget_edited'] = time();
		}

		return $instance;
	}

	/**
	 * Invalidate widget instance cache on Customizer save.
	 */
	public function update_widget() {

		$customized = filter_input( INPUT_POST, 'customized' );

		if ( empty( $customized ) ) {
			return;
		}

		$customized = json_decode( $customized, true );
		$changed    = false;

		foreach ( $customized as $key => $data ) {

			if ( ! isset( $data['encoded_serialized_instance'] ) || 0 !== stripos( $key, 'widget' ) ) {
				continue;
			}

			$instance                     = unserialize( base64_decode( $data['encoded_serialized_instance'] ) );
			$instance['fc_widget_edited'] = time();
			$instance                     = base64_encode( serialize( $instance ) );

			$data['encoded_serialized_instance'] = $instance;
			$data['instance_hash_key']           = wp_hash( $instance );
			$customized[ $key ]                  = $data;

			$changed = true;
		}

		if ( $changed ) {
			$_POST['customized'] = wp_json_encode( $customized );
		}
	}

	/**
	 * Set up and echo widget cache
	 *
	 * @param array  $instance Widget instance data.
	 * @param object $widget   Widget object instance.
	 * @param array  $args     Arguments.
	 *
	 * @return bool false
	 */
	public function widget_display_callback( $instance, $widget, $args ) {

		$edited = isset( $instance['fc_widget_edited'] ) ? $instance['fc_widget_edited'] : '';

		echo $this->fetch(
			$widget->id,
			array(
				'callback' => array( $widget, 'widget' ),
				'args'     => array( $args, $instance ),
			),
			$edited
		);

		return false;
	}

	/**
	 * Generate widget output, capture with buffer and timestamp.
	 *
	 * @param string $name Fragment name.
	 * @param array  $args Arguments.
	 *
	 * @return string
	 */
	protected function callback( $name, $args ) {

		ob_start();
		call_user_func_array( $args['callback'], $args['args'] );

		return ob_get_clean() . $this->get_comment( $name );
	}
}


php/class-gallery-cache.php
<?php

namespace Rarst\Fragment_Cache;

/**
 * Cache galleries.
 */
class Gallery_Cache extends Fragment_Cache {

	/** @var mixed $original_shortcode Callback originally registered for gallery shortcode. */
	protected $original_shortcode;

	/**
	 * @inheritDoc
	 */
	public function enable() {

		global $shortcode_tags;

		if ( isset( $shortcode_tags['gallery'] ) ) {
			$this->original_shortcode = $shortcode_tags['gallery'];
		}

		add_shortcode( 'gallery', array( $this, 'gallery_shortcode' ) );
	}

	/**
	 * @inheritDoc
	 */
	public function disable() {

		if ( ! empty( $this->original_shortcode ) ) {
			add_shortcode( 'gallery', $this->original_shortcode );
		}
	}

	/**
	 * Fetch and return cached gallery.
	 *
	 * @param array $args Gallery arguments.
	 *
	 * @return string
	 */
	public function gallery_shortcode( $args ) {

		if ( empty( $args ) ) {
			$args = array();
		}

		// Salt for cases post edited or attachments changed.
		$args['fc_post_modified']    = get_the_modified_time( 'U' );
		$args['fc_post_attachments'] = $this->get_attachment_ids();

		$post_id = get_the_ID();
		$output  = $this->fetch( 'post-' . $post_id, compact( 'args', 'post_id' ), $args );

		return $output;
	}

	/**
	 * Retrieve array of attachment IDs for current post.
	 *
	 * @return array
	 */
	public function get_attachment_ids() {

		static $attachments = array();

		$post_id = get_the_ID();

		if ( ! isset( $attachments[ $post_id ] ) ) {

			$attachments[ $post_id ] = get_posts( array(
				'post_type'   => 'attachment',
				'post_parent' => $post_id,
				'orderby'     => 'ID',
				'fields'      => 'ids',
			) );
		}

		return $attachments[ $post_id ];
	}

	/**
	 * Set up post context and generate gallery output.
	 *
	 * @param string $name Fragment name.
	 * @param array  $args Arguments.
	 *
	 * @return string
	 */
	protected function callback( $name, $args ) {

		global $post;

		$post = get_post( $args['post_id'] );
		setup_postdata( $post );
		$shortcode = isset( $this->original_shortcode ) ? $this->original_shortcode : 'gallery_shortcode';
		$output    = call_user_func( $shortcode, $args['args'] ) . $this->get_comment( $name );
		wp_reset_postdata();

		return $output;
	}
}


php/class-fragment-cache.php
<?php

namespace Rarst\Fragment_Cache;

/**
 * Abstract base class for implementation of fragment type handler.
 */
abstract class Fragment_Cache {

	/** @var bool $in_callback Static flag to catch and prevent nested caching calls.  */
	static $in_callback = false;

	/** @var int $timeout Cache timeout duration.  */
	public $timeout;

	/** @var string $type Handler type name. */
	protected $type;

	/**
	 * Create object and set parameters from passed.
	 *
	 * @param array $args Configuration arguments.
	 */
	public function __construct( $args ) {

		$this->type    = $args['type'];
		$this->timeout = $args['timeout'];
	}

	/**
	 * Enable fragment handler.
	 */
	abstract public function enable();

	/**
	 * Disable fragment handler.
	 */
	abstract public function disable();

	/**
	 * Wrapper to retrieve data through TLC Transient.
	 *
	 * @param string $name Name of fragment.
	 * @param array  $args Arguments.
	 * @param mixed  $salt Optional salt data.
	 *
	 * @return mixed
	 */
	public function fetch( $name, $args, $salt = '' ) {

		global $current_user;

		static $empty_user;

		if ( self::$in_callback || apply_filters( 'fc_skip_cache', false, $this->type, $name, $args, $salt ) ) {
			return $this->callback( $name, $args );
		}

		// Anonymize front-end run for consistency.
		if ( is_user_logged_in() ) {

			if ( empty( $empty_user ) ) {
				$empty_user = new \WP_User( 0 );
			}

			$stored_user  = $current_user;
			$current_user = $empty_user;
		}

		$salt   = maybe_serialize( $salt );
		$output = tlc_transient( apply_filters( 'fc_cache_name', 'fragment-cache-' . $this->type . '-' . $name . $salt, $this->type, $name, $args, $salt ) )
				->updates_with( array( $this, 'wrap_callback' ), array( $name, $args ) )
				->expires_in( $this->timeout )
				->get();

		if ( ! empty( $stored_user ) ) {
			$current_user = $stored_user;
		}

		return $output;
	}

	/**
	 * Wraps callback to correctly set execution flag.
	 *
	 * @param string $name Fragment name.
	 * @param array  $args Arguments.
	 *
	 * @return string
	 */
	public function wrap_callback( $name, $args ) {

		self::$in_callback = true;
		$output = $this->callback( $name, $args );
		self::$in_callback = false;

		return $output;
	}

	/**
	 * Used to generate data to be cached.
	 *
	 * @param string $name Fragment name.
	 * @param array  $args Arguments.
	 *
	 * @return string
	 */
	abstract protected function callback( $name, $args );

	/**
	 * Get human-readable HTML comment with timestamp to append to cached fragment.
	 *
	 * @param string $name Fragment name.
	 *
	 * @return string
	 */
	public function get_comment( $name ) {

		return '<!-- ' . esc_html( $name ) . ' ' . esc_html( $this->type ) . ' cached on ' . date_i18n( DATE_RSS ) . ' -->';
	}
}


php/class-menu-cache.php
<?php

namespace Rarst\Fragment_Cache;

/**
 * Cache navigation menus.
 */
class Menu_Cache extends Fragment_Cache {

	/**
	 * @inheritDoc
	 */
	public function enable() {

		global $wp_version;

		if ( is_admin() ) {
			add_action( 'admin_footer-nav-menus.php', array( $this, 'update_menus_edited' ) );
			add_action( 'wp_ajax_menu-locations-save', array( $this, 'update_menus_edited' ), 0 );
			add_action( 'wp_ajax_customize_save', array( $this, 'customize_save' ), 0 );

			return;
		}

		add_filter( 'pre_wp_nav_menu', array( $this, 'pre_wp_nav_menu' ), 10, 2 );
		add_filter( 'wp_nav_menu_objects', array( $this, 'wp_nav_menu_objects' ) );

		if ( version_compare( $wp_version, '3.9', '<' ) ) {
			add_filter( 'wp_nav_menu_args', array( $this, 'wp_nav_menu_args' ), 20 );
		}
	}

	/**
	 * @inheritDoc
	 */
	public function disable() {

		if ( is_admin() ) {
			remove_action( 'admin_footer-nav-menus.php', array( $this, 'update_menus_edited' ) );
			remove_action( 'wp_ajax_menu-locations-save', array( $this, 'update_menus_edited' ), 0 );
			remove_action( 'wp_ajax_customize_save', array( $this, 'customize_save' ), 0 );

			return;
		}

		remove_filter( 'pre_wp_nav_menu', array( $this, 'pre_wp_nav_menu' ), 10 );
		remove_filter( 'wp_nav_menu_objects', array( $this, 'wp_nav_menu_objects' ) );
		remove_filter( 'wp_nav_menu_args', array( $this, 'wp_nav_menu_args' ), 20 );
	}

	/**
	 * Return cached menu, using pre-generation hook.
	 *
	 * @param string $menu Menu HTML to return.
	 * @param object $args Menu arguments.
	 *
	 * @return string
	 */
	public function pre_wp_nav_menu( $menu, $args ) {

		$args                    = get_object_vars( $args );
		$args['echo']            = false;
		$args['fc_menus_edited'] = get_option( 'fc_menus_edited' );
		$name                    = is_object( $args['menu'] ) ? $args['menu']->slug : $args['menu'];

		if ( empty( $name ) && ! empty( $args['theme_location'] ) ) {
			$name = $args['theme_location'];
		}

		return $this->fetch( $name, $args, $args );
	}

	/**
	 * Fake no menu matches to force menu run custom callback.
	 *
	 * @deprecated
	 *
	 * @param array $args Menu arguments.
	 *
	 * @return array
	 */
	public function wp_nav_menu_args( $args ) {

		_deprecated_function( __FUNCTION__, '1.3', 'Menu cache with arguments override unnecessary on WP >= 3.9.' );

		if ( empty( $args['kessel_run'] ) ) {

			add_filter( 'wp_get_nav_menus', '__return_empty_array' ); // These are not the droids you are looking for.

			$args = array(
				'menu'           => '',
				'theme_location' => '',
				'fallback_cb'    => array( $this, 'fallback_cb' ),
				'original_args'  => $args,
			);
		}

		return $args;
	}

	/**
	 * Strip current* classes from menu items, since shared when cached.
	 *
	 * @param array $menu_items Array of menu item objects.
	 *
	 * @return array
	 */
	public function wp_nav_menu_objects( $menu_items ) {

		foreach ( $menu_items as $item_key => $item ) {
			foreach ( $item->classes as $class_key => $class ) {
				if ( 0 === stripos( $class, 'current' ) ) {
					unset( $menu_items[ $item_key ]->classes[ $class_key ] );
				}
			}
		}

		return $menu_items;
	}

	/**
	 * Save timestamp when menus were last modified for cache salt.
	 */
	public function update_menus_edited() {

		if ( ! empty( $_POST ) ) {
			update_option( 'fc_menus_edited', time() );
		}
	}

	/**
	 * Invalidate menu cache on related Customizer saves.
	 */
	public function customize_save() {

		$customized = filter_input( INPUT_POST, 'customized' );

		if ( empty( $customized ) ) {
			return;
		}

		$customized = json_decode( $customized, true );
		$settings   = array_keys( $customized );

		foreach ( $settings as $setting ) {

			if ( 0 === stripos( $setting, 'nav_menu' ) ) {

				update_option( 'fc_menus_edited', time() );

				return;
			}
		}
	}

	/**
	 * Restore arguments and fetch cached fragment for them.
	 *
	 * @deprecated
	 *
	 * @param array $args Arguments.
	 *
	 * @return string
	 */
	public function fallback_cb( $args ) {

		_deprecated_function( __FUNCTION__, '1.3', 'Menu cache with arguments override unnecessary on WP >= 3.9.' );

		remove_filter( 'wp_get_nav_menus', '__return_empty_array' );

		$args = $args['original_args'];
		unset( $args['original_args'] );
		$echo                    = $args['echo'];
		$args['echo']            = false;
		$args['kessel_run']      = true;
		$args['fc_menus_edited'] = get_option( 'fc_menus_edited' );
		$name                    = is_object( $args['menu'] ) ? $args['menu']->slug : $args['menu'];

		if ( empty( $name ) && ! empty( $args['theme_location'] ) ) {
			$name = $args['theme_location'];
		}

		$output = $this->fetch( $name, $args, $args );

		if ( $echo ) {
			echo $output;
		}

		return $output;
	}

	/**
	 * Generate and timestamp menu output.
	 *
	 * @param string $name Fragment name.
	 * @param array  $args Arguments.
	 *
	 * @return string
	 */
	protected function callback( $name, $args ) {

		remove_filter( 'pre_wp_nav_menu', array( $this, 'pre_wp_nav_menu' ), 10 );
		$output = wp_nav_menu( $args ) . $this->get_comment( $name );
		add_filter( 'pre_wp_nav_menu', array( $this, 'pre_wp_nav_menu' ), 10, 2 );

		return $output;
	}
}
